cpp/include/cudf/ast/detail/expression_evaluator.cuh

```diff
@@ -29,11 +29,31 @@
 
 #include <rmm/cuda_stream_view.hpp>
 
-namespace cudf {
+#include <cuda/std/type_traits>
+#include <cuda/std/utility>
 
-namespace ast {
+namespace cudf::ast::detail {
 
-namespace detail {
+CUDF_HOST_DEVICE constexpr bool is_complex_type(cudf::type_id type)
+{
+  return type == cudf::type_id::DECIMAL32 || type == cudf::type_id::DECIMAL64 ||
+         type == cudf::type_id::DECIMAL128 || type == cudf::type_id::STRING;
+}
+
+/**
+ * @brief Maps primitive types to their corresponding C++ types
+ *
+ * This template provides a mapping from cudf::type_id to concrete C++ types for
+ * common numeric types (integers, floating point, boolean) and temporal types
+ * (timestamps, durations).
+ *
+ * @tparam t The cudf::type_id to map to a C++ type
+ */
+template <cudf::type_id t>
+struct id_to_primitive_type {
```

Comment:
Typically we call `dispatch_void_if_*`. See `dispatch_void_if_nested` as an example.

Reply:
TIL `dispatch_void_if_*`

Reply:
Good point. This is a leftover since our initial idea was to dispatch all non-primitive types. Changed to `dispatch_void_if_complex`